src/sm.c

```diff
@@ -0,0 +1,480 @@
+            if (state1 != state2)
+            {
+                LogError("sm name=%s. state changed meanwhile from %" PRI_SM_STATE " to %" PRI_SM_STATE "", sm->name, SM_STATE_VALUE(state1), SM_STATE_VALUE(state2));
+                LONG n = InterlockedDecrement(&sm->n);
+                if (n == 0)
+                {
+                    WakeByAddressSingle((void*)&sm->n);
+                }
+                result = SM_EXEC_REFUSED;
+            }
+            else
+            {
+                /*Codes_SRS_SM_02_058: [ sm_exec_begin shall return SM_EXEC_GRANTED. ]*/
+                result = SM_EXEC_GRANTED;
+            }
+        }
+    }
+    return result;
+}
+
+void sm_exec_end(SM_HANDLE sm)
+{
+    /*Codes_SRS_SM_02_024: [ If sm is NULL then sm_exec_end shall return. ]*/
+    if (sm == NULL)
+    {
+        LogError("return");
+    }
+    else
+    {
+        LONG state = InterlockedAdd(&sm->state, 0);
+        if (
+            /*Codes_SRS_SM_02_059: [ If state is not SM_OPENED then sm_exec_end shall return. ]*/
+            ((state & SM_STATE_MASK) != SM_OPENED) &&
+            /*Codes_SRS_SM_02_060: [ If state is not SM_OPENED_DRAINING_TO_BARRIER then sm_exec_end shall return. ]*/
+            ((state & SM_STATE_MASK) != SM_OPENED_DRAINING_TO_BARRIER) &&
+            /*Codes_SRS_SM_02_061: [ If state is not SM_OPENED_DRAINING_TO_CLOSE then sm_exec_end shall return. ]*/
+            ((state & SM_STATE_MASK) != SM_OPENED_DRAINING_TO_CLOSE)
+            )
+        {
+            LogError("sm name=%s. cannot execute exec end when state is %" PRI_SM_STATE "", sm->name, SM_STATE_VALUE(state));
+        }
+        else
+        {
+            /*Codes_SRS_SM_02_062: [ sm_exec_end shall decrement n with saturation at 0. ]*/
+            do /*a rather convoluted loop to make _end be idempotent (too many _end will be ignored)*/
+            {
+                LONG n = InterlockedAdd(&sm->n, 0);
+                if (n <= 0)
+                {
+                    break;
+                }
+                else
+                {
+                    if (InterlockedCompareExchange(&sm->n, n - 1, n) != n)
+                    {
+                        /*well - retry sort of...*/
+                    }
+                    else
+                    {
+                        /*Codes_SRS_SM_02_063: [ If n reaches 0 then sm_exec_end shall signal that. ]*/
+                        if (n - 1 == 0)
+                        {
+                            WakeByAddressSingle((void*)&sm->n);
+                        }
+                        break;
+                    }
+                }
+            } while (1);
+        }
+    }
+}
+
+SM_RESULT sm_barrier_begin(SM_HANDLE sm)
+{
+    SM_RESULT result;
+    /*Codes_SRS_SM_02_027: [ If sm is NULL then sm_barrier_begin shall fail and return SM_ERROR. ]*/
+    if (sm == NULL)
+    {
+        LogError("return");
```

Comment:
>return [](start = 18, length = 6)

Also maybe NULL sm. return will be very ... generic if it shows up in logs

Reply:
The one that got away. Shall be fixed.

---
In reply to: [423416468](https://github.com/Azure/azure-c-util/pull/37#discussion_r423416468) [](ancestors = 423416468)

Reply:
Wow, turns out it is 3 of them! Fixed them all.

---
In reply to: [423422527](https://github.com/Azure/azure-c-util/pull/37#discussion_r423422527) [](ancestors = 423422527,423416468)